okhttp/src/main/java/okhttp3/internal/concurrent/TaskLogger.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.concurrent
+
+import java.util.logging.Level
+
+internal inline fun taskLog(
+  task: Task,
+  queue: TaskQueue,
+  messageBlock: () -> String
+) {
+  if (TaskRunner.logger.isLoggable(Level.FINE)) {
+    log(task, queue, messageBlock())
+  }
+}
+
+internal inline fun <T> logElapsed(
+  task: Task,
+  queue: TaskQueue,
+  block: () -> T
+): T {
+  var startNs = -1L
+  val loggingEnabled = TaskRunner.logger.isLoggable(Level.FINE)
+  if (loggingEnabled) {
+    startNs = queue.taskRunner.backend.nanoTime()
+    log(task, queue, "starting")
+  }
+
+  var completedNormally = false
+  try {
+    val result = block()
+    completedNormally = true
+    return result
+  } finally {
+    if (loggingEnabled) {
+      val elapsedNs = queue.taskRunner.backend.nanoTime() - startNs
+      if (completedNormally) {
+        log(task, queue, "finished in ${formatTime(elapsedNs)}")
+      } else {
+        log(task, queue, "failed in ${formatTime(elapsedNs)}")
```

Comment:
worth including exn.toString() in the message?

Reply:
This is a case where we’re going to re-throw. Don’t want this logging to be any kind of substitute for application-layer troubleshooting.